ntbs-service/DataMigration/NotificationMapper.cs

```diff
@@ -37,49 +37,94 @@ public NotificationMapper(IMigrationRepository migrationRepository, IReferenceDa
+                var manualTestResults = _migrationRepository.GetManualTestResults(legacyIds);
+                var socialContextVenues = _migrationRepository.GetSocialContextVenues(legacyIds);
+                var socialContextAddresses = _migrationRepository.GetSocialContextAddresses(legacyIds);
+                var transferEvents =  _migrationRepository.GetTransferEvents(legacyIds);
+                // TODO outcome events
 
-            return Task.WhenAll(notificationGroups
-                .Select(group =>
-                {
-                    return Task.WhenAll(group.Select(AsNotificationAsync));
-                })).Result;
+                return await CombineDataForGroup(
+                    legacyIds,
+                    (await legacyNotifications).ToList(),
+                    (await sitesOfDisease).ToList(),
+                    (await manualTestResults).ToList(),
+                    (await socialContextVenues).ToList(),
+                    (await socialContextAddresses).ToList(),
+                    (await transferEvents).ToList()
+                    );
+            }));
         }
 
-        private async Task<IEnumerable<IGrouping<string, dynamic>>> GetDiseaseSitesPerNotification(IEnumerable<dynamic> notificationsRaw)
+        private async Task<IEnumerable<Notification>> CombineDataForGroup(List<string> legacyIds,
+            IList<dynamic> notifications,
+            IList<dynamic> sitesOfDisease,
+            IList<dynamic> manualTestResults,
+            IList<dynamic> socialContextVenues,
+            IList<dynamic> socialContextAddresses, 
+            IList<dynamic> transferEvents)
         {
-            var legacyIds = notificationsRaw.Select(x => x.OldNotificationId).Cast<string>();
-            var notificationSitesRaw = await _migrationRepository.GetNotificationSites(legacyIds);
-            return notificationSitesRaw.GroupBy(x => x.OldNotificationId as string);
+            return await Task.WhenAll(legacyIds.Select(async id =>
+            {
+                var rawNotification = notifications.Single(n => n.OldNotificationId == id);
+                var notificationSites = AsSites(sitesOfDisease
+                    .Where(s => s.OldNotificationId == id)
+                );
+                var notificationTestResults = manualTestResults
+                    .Where(t => t.OldNotificationId == id)
+                    .Select(AsManualTestResult)
+                    .ToList();
+                var notificationSocialContextVenues = socialContextVenues
+                    .Where(sc => sc.OldNotificationId == id)
+                    .Select(AsSocialContextVenue)
+                    .ToList();
+                var notificationSocialContextAddresses = socialContextAddresses
+                    .Where(sc => sc.OldNotificationId == id)
+                    .Select(AsSocialContextAddress)
+                    .ToList();
+                var notificationTransferEvents = await Task.WhenAll(transferEvents
```

Comment:
Why are we threading this (is it too slow normally?)

Reply:
`AsTransferEvent` is `async` since it calls the DB (to map the hospital to the service)